Improve clarity of the handle example (#248)
ietf-rats-wg/architecture: draft-ietf-rats-architecture.md

## Example 3: Handle-based Passport Model Example

> A third option to establish the epoch of Evidence or Attestation Result
> generation - next to nonces and timestamps ({{freshness}}) - is by using
> "handles".

Reviewer: ```suggestion
The following example illustrates a hypothetical Passport Model
solution that uses handles instead of timestamps or nonces.
```
For consistency with the start of examples 1 and 2

> Handles are opaque, non-predictable data regularly broadcast to protocol
> participants by a Handle Distributor.  Each handle represents a specific epoch,

Reviewer 2: ```suggestion
Handles are opaque, non-predictable data regularly broadcast by a Handle Distributor.  
Each handle represents a specific epoch,
```

> in which generated Evidence can be asserted to be fresh.

Reviewer: The freshness section still calls them nonces in lines 1109-1114. We should be consistent, maybe by calling them handles there.   And these two sentences are not about the example per se, they're about the definition of handles and so I think they should move to where line 1109-1114 are.

Reviewer 3: Dave took on the task to generate text in the Freshness section that introduces the Handle concept in the body of the document.

Reviewer: Recommend rebasing on top of PR #257

The distribution of a new handle determines the transition from one epoch to
the next.  The transition between epochs goes through an "overlap" phase in
which the previous epoch is still extant while the current epoch is
established.  The span of the overlap phase is determined solely by the
transmission delay of the underlying communication channel,

> `time(HR)-time(HT)`.  Once the overlap phase is over, an epoch is fully

Reviewer: I don't think this is true, since the receiver has no idea when time(HT) is.   To avoid race conditions, receivers (Verifiers / Relying Parties) need to accept both the previous handle and the current handle.  In that case, the original text inequalities are correct and this one is not.

Author: > I don't think this is true, since the receiver has no idea when time(HT) is. 

I see your point, but I wouldn't say it is a false statement.  It is however both imprecise (it should be the worst case delay among all receivers) and not really useful from an operational point of view.  Therefore I'm happy to remove it from the sentence which has clear enough prose to express the idea already.

I think we also need to highlight that the extent of the overlap, the thing previously known as "offset" (would "pad" be a better name?), is a central protocol parameter which needs calibration and agreement among all the participants to be meaningful.  It's not a unilateral decision that a protocol participant makes up by itself.

And another related thing (which I just realised, sorry for overloading the comment) is that the discussion seems to be assuming a network topology where V is _closer_ to the Handle Distributor than V - and therefore V’s epoch starts _before_ A’s epoch.  We should also consider the dual situation and therefore the case where Evidence arrives at V with a handle that V has not yet received.  This calls for extending the “offset” on the left to buffer _maybe valid_ handles?

> [...] the original text inequalities are correct and this one is not

One of the troubles I was having with the original text is that I really don’t understand what the inequality is supposed to say.  In particular, how is `time(HR_v)-time(EG_a)` a definition of “age of the handle used”?  Besides, it seems that there is an unstated assumption here that A and V have synchronised clocks - otherwise time(EG_a) and time(HR_v) are not on the same timeline and can't be meaningfully combined.

Reviewer 3: First of all, you intended to write "where V is _closer_ to the Handle Distributor than A", I think.

Additionally, I think we came to the conclusion that it is preferable to use an event (defined in the table) in the text of the Appendix in order to provide better context. Simply removing the sentence with time(HT) at the end would go against that. Hence, I would recommend to not simply remove it, but to replace it with a paragraph that highlights why event types, such as time(HT), that can frame the extent of an overlap phase are vital to readers of the document here.

In general, Handle Distribution is in support of various scenarios. In some of them the receivers have no clocks at all, in some of them only a few - e.g. the Verifier. Maybe it would help with the scenario description, if we add a Clock to the Verfier? (so that only the swarm of Attesters has no clock)

Author: > First of all, you intended to write "where V is _closer_ to the Handle Distributor than A", I think.

Your thinking is correct, fixed.  Thanks for spotting it!

> Additionally, I think we came to the conclusion that it is preferable to use an event (defined in the table) in the text of the Appendix in order to provide better context. Simply removing the sentence with time(HT) at the end would go against that. Hence, I would recommend to not simply remove it, but to replace it with a paragraph that highlights why event types, such as time(HT), that can frame the extent of an overlap phase are vital to readers of the document here.

I agree.  Frankly, I care more about clarity at the conceptual level rather than anything else.  So, whatever device helps to better grasp the ideas is very welcome.

> In general, Handle Distribution is in support of various scenarios. In some of them the receivers have no clocks at all, in some of them only a few - e.g. the Verifier. Maybe it would help with the scenario description, if we add a Clock to the Verfier? (so that only the swarm of Attesters has no clock)

What I was trying to say is just that the second inequality makes no sense because the two `time()` functions apply to different timelines (and are therefore the comparison is between pears and apples!) unless the underlying clocks at V and A are synchronised.  If we cannot assume the presence of clocks at nodes implementing the A role, let alone synchronised clocks, well that's another good reason for dropping the inequality, isn't it?

If instead there is at least some assumption on clock presence in some of the roles (and possibly a synchronisation requirement?), that should be clearly stated.

> `time(HR)-time(HT)`.  Once the overlap phase is over, an epoch is fully

Reviewer: I don't think this is correct.  The overlap is `time(HR')-time(HR)`.

established.

In order to determine freshness of Evidence or Attestation Results, consumers
of messages based on handles need to understand whether an epoch is fully

> established or in overlap phase.

Reviewer: This is confusing.  Remove it?

Author: Can you please explain where is the confusion?

We think that being able to distinguish between a fully established epoch and one that is still in overlap phase is crucial for a receiver that needs to decide whether to accept or refuse an incoming message.

Reviewer: I don't understand your point.   If it's in overlap phase, it should still be accepted.

Author: Sure.  However, in order to be able to _reject_ a message coming from the previous epoch a receiver needs to be able to say: _"overlap is over and the new epoch is now fully established, therefore I will assume whatever comes from a previous epoch is a replay"_

Reviewer 3: Accepting an incoming Conceptual Message without overlap phase awareness should be very strongly discouraged, as it can break the whole RATS concept. I understand that it is not obvious or easy for a receiver to infer that it is in an overlap phase. Nevertheless, delimiting factors should be illustrated in this context as best as we can.

Reviewer 3: It was pointed out to me that my comment was mislead. Unfortunately, I have to agree.

What I intended to convey was: yes, of course the message is accepted, but blindly accepting it without awareness of an overlap phase poses a security risk in the following appraisal procedures due to lack of relevant context information. IHTH

> The Handle Distributor, hosted in a dedicated entity or co-located with other
> RATS roles, is trusted by the protocol participants to behave correctly with

Reviewer: ```suggestion
The Handle Distributor, hosted in a dedicated entity or co-located with another
RATS role, is trusted by the protocol participants to behave correctly with
```

Reviewer 3: Why couldn't it be collapsed with more than one other RATS role?

Reviewer 4: "... co-located with other RATS roles, is trusted..."

regards to its duty of handle distribution.

Consumers of Conceptual Messages based on handles must be able to safely trace
each handle's origin back to the broadcasting Handle Distributor either using
topological considerations (e.g., co-location), or by cryptographic means.

The example in {{fig-handle-passport}} illustrates a Handle-based interaction

> using the passport topology model.

Reviewer: This is either redundant with the first sentence of this section, or everything above here should be moved to the freshness section.

> At `time(HT)`, the Handle Distributor broadcasts handle `H` which starts a new

Reviewer: It actually doesn't matter whether it's "broadcast" or only unicast to two specific entities (e.g., an Attester and Verifier) that use it.
```suggestion
At `time(HT)`, the Handle Distributor sends handle `H` which starts a new
```

> epoch `E` for a protocol participant upon reception at `time(HR)`.

Reviewer: `E` in line 1492 is Evidence whereas here it's Epoch, which is confusing.   I'd like to fix line 1492 to not abbreviate Evidence

The Attester generates Evidence incorporating handle `H` and conveys it to the
Verifier.

The Verifier appraises that the received handle `H` is "fresh" according to the

> definition provided above, and generates an Attestation Result.  The

Reviewer: which line is "definition provided above" referring to?

Attestation Result is conveyed to the Attester.

After the transmission of handle `H'` at `time(HT')` a new epoch `E'` is
established when `H'` is received by each protocol participant.  The Attester
relays the Attestation Result obtained during epoch `E` (associated with handle

> `H`) to the Relying Party using the handle for the current epoch `H'`.

Reviewer: and what happens if the Relying Party hasn't gotten H' yet?   (I added a brief discussion of this in PR 257, so would be good to clarify here.  This section just has an example so any answer may be ok.)

In the illustrated scenario, the Handle for relaying an Attestation Result to

> the Relying Party is fresh, while the handle that was used to generate the

Reviewer 4: The word 'while' can be removed or changed to "but" or "however"

> handle is stale. This indicates that at least one epoch has passed in between,

Reviewer: "handle used to generate the handle" is confusing since no such concept was introduced earlier

Author: I think it should say "the handle [...] used to generate the Evidence is stale" instead?

> handle is stale. This indicates that at least one epoch has passed in between,

Reviewer 3: ```suggestion
Evidence is stale. This indicates that at least one epoch has passed in between,
```
That's a typo. Evidence is correct.

        |                |                |               |
        ~                ~                ~               ~
        |                |                |               |

>      time(HR_a)<----time(HT)-----------time(HR_v)----->time(HR_r)

Reviewer: time(HT) is not used in any computation or inequality so should be removed for consistency

Reviewer 3: I think the point was to provide more context by using a defined event in the text. I am rather cautious to agree with the notion that "used in a computation or inequality" is a needed requirement here. Also, it is used in a comparison/computation in the text.

We all agree that, for example, illustrating time(HR_a) in the diagram - although it cannot be captured by an Attester with no clock - is okay. Consequently, highlighting that time(HT) and time(HR_a) are related to effectively existing overlap phases although the delta is not easy to capture seems fine to me.

Reviewer: I agree that illustrating time(HR_a) in the diagram is okay so long as it's used in a message or computation/inequality.

        |                |                |               |
     time(EG_a)          |                |               |

>         |--E{H,time(VG_a)}--------------->|               |

Reviewer: Problems:
1) the term E is not defined, only Evidence
2) You cannot convey time(VG_a) without synchronized clocks
3) time(EG_a) is left in line 1491 but then not used

Reviewer 3: 1.) As far as I understand it, this seems to be a proposal simplify the layout of the diagram.

The current state evolved into a representation that was neither readable nor compressible:
  
```
     time(EG_a)
        |----Evidence{time(EG_a)}
        | {Handle1,time(EG_a)-time(VG_a)}
```
Whatever the solution in the end, this has to be fixed, I think (the { vs ( syntax is never introduced and not intuitive).

2. true, we have to walk through the sequence again together, I think

3. It still happens. Is it okay to not illustrates what happens, but is not conveyed in a Conceptual Message? I have some doubts that "used" only is restricted to conveyance or am I missing something obvious here?

Reviewer: In the snippet in Henk's comment there's two {} blocks on the same message, which should never happen.
{} means "includes (inside)".  That is Evidence{time(EG_a)} means that time(EG_a) is inside the signed Evidence.   I completely agree it would be good to explain the notation before or in the first example where it's used.

        |                |                |               |
        |                |             time(RG_v)         |

>         |<----------------R{H,time(RX_v)}-|               |

Reviewer: 1) R is undefined
2) time(RX_v) is not useful to convey without synchronized clocks.  There's nothing the RP can do with it in this section.

        |                |                |               |
     time(HR'_a)<---time(HT')----------time(HR'_v)---->time(HR'_r)
        |                |                |               |
     time(RR_a)          |                |               |

>         |--RR{H',R{H,time(RX_v)}}--------------------->time(RA_r)

Reviewer 2: ```suggestion
        |     Attestation Result
        |-------{H',R{H,time(RX_v)}}------------------>time(RA_r)
```

Reviewer 3: ```suggestion
        |---R{H',R{H,time(RX_v)}}--------------------->time(RA_r)
```

We need a legend. 'R' is short for Attestation Result.

The point of the illustration is that a Conceptual message with a stale handle is conveyed and that the Relying Party can detect that the Handle used in the Conceptual Message -- the relayed Attestation Result -- is not the current handle anymore.

Instead of adding Attestation Result text in the diagram, I simply removed the RR --> which renders the Relayed Attestation Result message a Attestation Message. This probably could be good enough, if the English text than captures that context.

Reviewer 2: I think that you should leave the text in the diagram fully expanded until we can agree that the new HT math is correct.
Then, let's see if we can find better shortcuts/abbreviations.

Reviewer 4: I agree there should be a legend. 
However, these examples are very complex and very terse for an architecture document. It would be better to IMO to number each message in the sequence diagram flows then explain them in more readable detail in a numbered prose section below the diagram.

## Example 3: Handle-based Passport Model Example

The example in {{fig-handle-passport}} illustrates a hypothetical Passport Model
solution that uses handles instead of nonces or timestamps.

The Handle Distributor broadcasts handle `H` which starts a new
epoch `E` for a protocol participant upon reception at `time(HR)`.

The Attester generates Evidence incorporating handle `H` and conveys it to the
Verifier.

The Verifier appraises that the received handle `H` is "fresh" according to the

> definition provided in {{freshness}}, and generates an Attestation Result.  The

Reviewer 2: ```suggestion
definition provided in {{epochfreshness}}, and generates an Attestation Result.  The
```

Attestation Result is conveyed to the Attester.

After the transmission of handle `H'` a new epoch `E'` is
established when `H'` is received by each protocol participant.  The Attester
relays the Attestation Result obtained during epoch `E` (associated with handle

> `H`) to the Relying Party using the handle for the current epoch `H'`.

Reviewer 2: ```suggestion
`H`) to the Relying Party using the handle for the current epoch `H'`.
If the Relying Party had not yet received `H'`, then the Attestation Result would be rejected, but in this example, it is received.
```

The example in {{fig-handle-passport}} illustrates a Handle-based interaction
using the passport topology model.

The Handle Distributor broadcasts handle `H` which starts a new
epoch `E` for a protocol participant upon reception at `time(HR)`.

The Attester generates Evidence incorporating handle `H` and conveys it to the
Verifier.

The Verifier appraises that the received handle `H` is "fresh" according to the
definition provided in {{freshness}}, and generates an Attestation Result.  The
Attestation Result is conveyed to the Attester.

After the transmission of handle `H'` a new epoch `E'` is
established when `H'` is received by each protocol participant.  The Attester
relays the Attestation Result obtained during epoch `E` (associated with handle
`H`) to the Relying Party using the handle for the current epoch `H'`.

In the illustrated scenario, the handle for relaying an Attestation Result to the Relying Party is current, while a previous handle was used to generate Verifier evaluated evidence. 
This indicates that at least one epoch transition has occurred, and the Attestation Results may only be as fresh as the previous epoch.

> which can be taken into account in Appraisal Policies for Attestation Results
> about freshness.

Reviewer 2: ```suggestion

```

## Example 3: Handle-based Passport Model Example

> The example in {{fig-handle-passport}} illustrates a Handle-based interaction
> using the passport topology model.

Reviewer: ```suggestion
The example in {{fig-handle-passport}} illustrates a hypothetical Passport Model
solution that uses handles instead of nonces or timestamps.
```
Keep wording consistent with intros to the previous sections.

The Handle Distributor broadcasts handle `H` which starts a new
epoch `E` for a protocol participant upon reception at `time(HR)`.

The Attester generates Evidence incorporating handle `H` and conveys it to the
Verifier.

The Verifier appraises that the received handle `H` is "fresh" according to the

> definition provided in {{freshness}}, and generates an Attestation Result.  The

Reviewer: I'd recommend referencing the "Implicit Timekeeping using Epoch Handles" subsection rather than the full "Freshness" section.  And "according to the definition provided" there is ambiguous since it mentions that any solution has to take one of three approaches, so need to clarify which approach is shown in this example.

I.e., quoting text there:
> An actual solution also needs to take into account race conditions
> when transitioning to a new epoch, such as by using a counter signed
> by the Handle Distributor as the handle, or by including both the current and
> previous handles in messages and/or checks, by requiring retries
> in case of mismatching handles, or by buffering incoming messages
> that might be associated with a handle that the receiver has not yet
> obtained.

Attestation Result is conveyed to the Attester.

After the transmission of handle `H'` a new epoch `E'` is
established when `H'` is received by each protocol participant.  The Attester
relays the Attestation Result obtained during epoch `E` (associated with handle
`H`) to the Relying Party using the handle for the current epoch `H'`.

> In the illustrated scenario, the Handle for relaying an Attestation Result to
> the Relying Party is fresh, while the handle that was used to generate the
> handle is stale. This indicates that at least one epoch has passed in between,

Reviewer 2: ```suggestion
In the illustrated scenario, the handle for relaying an Attestation Result to the Relying Party is current, while a previous handle was used to generate Verifier evaluated evidence. 
This indicates that at least one epoch transition has occurred, and the Attestation Results may only be as fresh as the previous epoch.
```

        |                |                |               |
        ~                ~                ~               ~
        |                |                |               |
     time(HR_a)<---------+-----------time(HR_v)----->time(HR_r)
        |                |                |               |
     time(EG_a)          |                |               |

>         |---Evidence{H,time(VG_a)}------->|               |

Reviewer 2: ```suggestion
        |---Evidence{H,time(EG_a)-time(VG_a)}----->|               |
```

        |                |                |               |
        |                |           time(RG_v)           |
        |<--Attestation Result------------|               |

>         |   {H,time(RX_v)}                |               |

Reviewer 2: ```suggestion
        |   {H,time(RX_v)-time(RG_v)}                |               |
```

        |                |                |               |
     time(HR'_a)<--------+-----------time(HR'_v)---->time(HR'_r)
        |                |                |               |
     time(RR_a)          |                |               |
        |---Attestation Result---------------------->time(RA_r)

>         |   {H',R{H,time(RX_v)}}          |               |

Reviewer 2: ```suggestion
        |   {H',R{H,time(RX_v)-time(RG_v)}}          |               |
```

        |                |                |               |
        ~                ~                ~               ~
        |                |                |               |
     time(HR_a)<---------+-----------time(HR_v)----->time(HR_r)
        |                |                |               |
     time(EG_a)          |                |               |

>         |---Evidence{H,time(VG_a)}------->|               |

Reviewer: time(VG_a) makes no sense in this example since clocks are not synchronized so the actual value is meaningless, and there's no other time(foo_a) transmitted to the attester to use as a delta.
The old (red) text had time(EG_a)-time(VG_a) which did make sense, as the delta is meaningful to the receiver (it just requires a clock in the Attester that need not be synchronized).

That's why the old text had time(RX_v)-time(RG_v) is is meaningful.

        |                |                |               |
        |                |           time(RG_v)           |
        |<--Attestation Result------------|               |

>         |   {H,time(RX_v)}                |               |

Reviewer: time(RX_v) is useless in this message for the same reason as in my previous comment.
